### PR TITLE
Remove BinaryFormatter from StateFileBase

### PR DESCRIPTION
**Pull request by Forgind**
This ends our reliance on BinaryFormatter within any StateFileBase. This comprises AssemblyReferenceCache, ResolveComReference, and ResGenDependencies. SystemState (part of ResolveAssemblyReference) also extends StateFileBase, but that was already converted. We may want to combine those two serialization efforts into a single location, but that's more of a refactor so lower priority.

Also added tests.

### CODE REVIEW

File: src/Tasks/ResGenDependencies.cs [hunk 1]
```diff
@@ -284,9 +286,13 @@ private static string[] GetLinkedFiles(string filename, string baseLinkedFileDir
         [Serializable]
```
Comment (rainersigwald) [inline src/Tasks/ResGenDependencies.cs:286]: Getting rid of this too?

File: src/Tasks/TaskTranslatorHelpers.cs [hunk 2]
```diff
@@ -34,5 +35,94 @@ public static void Translate(this ITranslator translator, ref FrameworkName fram
...
+                {
+                    string key = kvp.Key;
+                    DateTime val = kvp.Value;
+                    translator.Translate(ref key);
+                    translator.Translate(ref val);
+                }
+            }
+        }
+
+        public static void Translate(this ITranslator translator, ref Dependencies dependencies, Type t)
```
Comment (rainersigwald) [inline src/Tasks/TaskTranslatorHelpers.cs:69]: Why are you putting these here instead of as a method in the relevant classes?

Comment (Forgind) [reply src/Tasks/TaskTranslatorHelpers.cs:69]: When we've added new Translators, we've generally done it in "TranslatorHelpers" (or similar) files. I was thinking about eventually combining these with @rokonec's change for the other kind of StateFileBase: https://github.com/dotnet/msbuild/pull/6094

TaskTranslatorHelpers makes a little more sense to me than a shared TranslatorHelpers, though, because these are all task-specific.

File: src/Tasks/ResGenDependencies.cs (hunk 1 above)
Comment (Forgind) [reply src/Tasks/ResGenDependencies.cs:286]: Yep! Thanks.

File: src/Tasks/TaskTranslatorHelpers.cs (hunk 2 above)
Comment (rokonec) [reply src/Tasks/TaskTranslatorHelpers.cs:69]: I believe that for for our own types we shall implement ITranslatable interface and use extension only for close source clases, like Dictionary<X,Y> or Version. 
This was the serialization code stays inside the class, which might limit futuru serialization code omission bugs.

File: src/Tasks/StateFileBase.cs [hunk 3]
```diff
@@ -75,43 +72,30 @@ internal static StateFileBase DeserializeCache(string stateFile, TaskLoggingHelp
...
-                        retVal = deserializedObject as StateFileBase;
+                        var translator = BinaryTranslator.GetReadTranslator(s, buffer: null);
+                        translator.Translate(ref retVal, requiredReturnType);
 
-                        // If the deserialized object is null then there would be no cast error but retVal would still be null
-                        // only log the message if there would have been a cast error
-                        if (retVal == null && deserializedObject != null)
+                        // If retVal is still null or the version is wrong, log a message not a warning. This could be a valid cache with the wrong version preventing correct deserialization.
+                        // For the latter case, internals may be unexpectedly null.
+                        if (retVal == null || retVal._serializedVersion != CurrentSerializationVersion)
```
Comment (rokonec) [inline src/Tasks/StateFileBase.cs:80]: Since _serializedVersion is never serialized, it will **always** have same value as CurrentSerializationVersion and it will never detect outdated cache.
In SystemState.DeserializeCacheByTranslator I have implemented some kind of file signature check along with basic versioning.

File: src/Tasks/TaskTranslatorHelpers.cs [hunk 4]
```diff
@@ -34,5 +35,94 @@ public static void Translate(this ITranslator translator, ref FrameworkName fram
...
+            {
+                ResGenDependencies rgd = stateFile as ResGenDependencies;
+                rgd ??= new();
+                translator.Translate(ref rgd.resXFiles, typeof(ResGenDependencies.ResXFile));
+                translator.Translate(ref rgd.portableLibraries, typeof(ResGenDependencies.PortableLibraryFile));
+                translator.Translate(ref rgd.baseLinkedFileDirectory);
+                stateFile = rgd;
+            }
+#if NETFRAMEWORK
+            else if (t == typeof(ResolveComReferenceCache))
```
Comment (rokonec) [inline src/Tasks/TaskTranslatorHelpers.cs:108]: When called in dotnet core it will write nothing into file. If such file will be then deserialized in NETFRAMEWORK it will crash.
How do we protect against cache mangling between CLI and VS msbuild invocations?
I have not looked at caller side to verify that for given types it is never called for non NETFRAMEWORK, but even if it is assured, it makes code somehow fragile.

**Review by rokonec — CHANGES_REQUESTED** [review]

File: src/Tasks/TaskTranslatorHelpers.cs (hunk 4 above)
Comment (Forgind) [reply src/Tasks/TaskTranslatorHelpers.cs:108]: It should only be .NET Framework --> .NET Framework. Those caches are only compiled, even, if the TFI is .NETFramework:
https://github.com/dotnet/msbuild/blob/main/src/Tasks/Microsoft.Build.Tasks.csproj#L558

If, at some point, we were to add support for those on .NET Core, we'd have to start by moving that to a not-framework-specific itemgroup, which would make the tests I wrote (and didn't specify not-core on) fail.

File: src/Tasks/StateFileBase.cs (hunk 3 above)
Comment (Forgind) [reply src/Tasks/StateFileBase.cs:80]: Good point, added.

File: src/Tasks/TaskTranslatorHelpers.cs [hunk 5]
```diff
@@ -34,5 +35,35 @@ public static void Translate(this ITranslator translator, ref FrameworkName fram
...
+
+        public static void TranslateDictionary(this ITranslator translator, ref Dictionary<string, DateTime> dict, StringComparer comparer)
+        {
+            int count = 0;
+            if (translator.Mode == TranslationDirection.ReadFromStream)
+            {
+                dict = new Dictionary<string, DateTime>(comparer);
+                translator.Translate(ref count);
+                string key = string.Empty;
+                DateTime val = DateTime.Now;
```
Comment (rainersigwald) [inline src/Tasks/TaskTranslatorHelpers.cs:47]: Would it be harmful to default this to min or max? Those are constants and would never require a clock read.

```suggestion
                DateTime val = DateTime.MinValue;
```

File: src/Tasks/TaskTranslatorHelpers.cs [hunk 6]
```diff
@@ -34,5 +35,35 @@ public static void Translate(this ITranslator translator, ref FrameworkName fram
                 frameworkName = new FrameworkName(identifier, version, profile);
             }
         }
+
+        public static void TranslateDictionary(this ITranslator translator, ref Dictionary<string, DateTime> dict, StringComparer comparer)
```
Comment (rainersigwald) [inline src/Tasks/TaskTranslatorHelpers.cs:39]: Why is this overload not in `BinaryTranslator.cs` with the rest? It doesn't use a Tasks-specific type.

File: src/Tasks/SystemState.cs [hunk 7]
```diff
@@ -329,6 +246,7 @@ public void Translate(ITranslator translator)
                 ref instanceLocalFileStateCache,
                 StringComparer.OrdinalIgnoreCase,
                 (ITranslator t) => new FileState(t));
+            IsDirty = false;
```
Comment (rainersigwald) [inline src/Tasks/SystemState.cs:249]: Is this correct for both serialization and deserialization directions?

File: src/Tasks/SystemState.cs [hunk 8]
```diff
@@ -89,11 +84,6 @@ internal sealed class SystemState : StateFileBase, ITranslatable
         /// </summary>
         private GetAssemblyMetadata getAssemblyMetadata;
 
-        /// <summary>
-        /// Cached delegate.
-        /// </summary>
-        private FileExists fileExists;
```
Comment (rainersigwald) [inline src/Tasks/SystemState.cs:95]: What's the motivation for this change?

File: src/Tasks/Dependencies.cs [hunk 9]
```diff
@@ -1,32 +1,65 @@
...
+        internal Dependencies(ITranslator translator, Type t)
+        {
+            Translate(translator, t);
+        }
+
+        public void Translate(ITranslator translator, Type t)
+        {
+            translator.TranslateDictionary(ref dependencies, (ITranslator translator, ref DependencyFile dependency) =>
+            {
+                if (t == typeof(ResGenDependencies.ResXFile))
```
Comment (rainersigwald) [inline src/Tasks/Dependencies.cs:32]: Why pass and check the type explicitly? Could you instead do a pattern match on the runtime type?

File: src/Tasks/ResGenDependencies.cs [hunk 10]
```diff
@@ -90,11 +89,25 @@ internal bool UseSourcePath
...
+            translator.Translate(ref baseLinkedFileDirectory);
+            translator.Translate(ref _serializedVersion);
+        }
+
         internal ResXFile GetResXFileInfo(string resxFile, bool useMSBuildResXReader)
         {
             // First, try to retrieve the resx information from our hashtable.
-            var retVal = (ResXFile)resXFiles.GetDependencyFile(resxFile);
-            if (retVal == null)
+            if (resXFiles.GetDependencyFile(resxFile) is not ResXFile retVal || retVal == null)
```
Comment (rainersigwald) [inline src/Tasks/ResGenDependencies.cs:110]: I don't feel like this is clearer than switching the cast to an `as`.

File: src/Tasks/TaskTranslatorHelpers.cs (hunk 5 above)
Comment (Forgind) [reply src/Tasks/TaskTranslatorHelpers.cs:47]: Nope, just needed a value.

File: src/Tasks/SystemState.cs (hunk 7 above)
Comment (Forgind) [reply src/Tasks/SystemState.cs:249]: Yes. For serialization, the cache was just written to disk, so it is no longer dirty. For deserialization, this cache was just read from disk, so it is no longer dirty. Either way, the two versions match.

File: src/Tasks/SystemState.cs (hunk 8 above)
Comment (Forgind) [reply src/Tasks/SystemState.cs:95]: It isn't used, and I came across it. Small cleanup.

File: src/Tasks/SystemState.cs (hunk 7 above)
Comment (rainersigwald) [reply src/Tasks/SystemState.cs:249]: Great! Can you comment that please?

File: src/Tasks/Dependencies.cs (hunk 9 above)
Comment (Forgind) [reply src/Tasks/Dependencies.cs:32]: Great question—that's actually how I first implemented it, and it took me a bit to figure out why it was failing. For serializing to disk, yes, that works. For deserializing, dependency comes in as null, so it can be cast to either a ResXFile or a PortableLibraryFile. Only one of those is valid, however, so I need to pick the right one. Explicitly passing the type (or some equivalent information) is the only way I can get the requisite information for deserializing the right type.

File: src/Tasks/ResGenDependencies.cs (hunk 10 above)
Comment (Forgind) [reply src/Tasks/ResGenDependencies.cs:110]: That's probably better, thanks. This wasn't about clarity but correctness—the cast was failing and throwing an exception because it wasn't a ResXFile. That shouldn't happen in practice, and it's now fixed (see previous comment), but did cause problems for a bit.

File: src/Tasks/TaskTranslatorHelpers.cs (hunk 6 above)
Comment (Forgind) [reply src/Tasks/TaskTranslatorHelpers.cs:39]: To do this, I would have to add a method signature in ITranslator.cs and a separate implementation for read and write. Those implementations would be almost identical and easily merged. Looking at other examples in that class, most just reimplemented the same logic twice with a tiny tweak. I don't approve of wasting code like that and would propose moving all methods beyond the basics (int, string, etc.) to an extensions class where they can reuse code properly.

File: src/Tasks/Dependencies.cs (hunk 9 above)
Comment (rokonec) [reply src/Tasks/Dependencies.cs:32]: Can we make Dependencies implement ITranslatable and override it in derived classes?

Comment (rainersigwald) [reply src/Tasks/Dependencies.cs:32]: If it was the first instinct for both of us it deserves a comment please

Comment (Forgind) [reply src/Tasks/Dependencies.cs:32]: @rokonec The type is just Dictionary<string, DependencyFile> and it has to work for either Dictionary<string, ResXFile> or Dictionary<string, PortableLibraryFile>. The Dependencies class doesn't know which will be required, so the Translator can't either. To be honest, this whole class structure could have been dramatically simplified by just having ResGenDependencies have a Dictionary<string, ResXFile> and a Dictionary<string, PortableLibraryFile>. That would make this a lot cleaner but be a little outside the scope of this PR. That said, if you want me to do that, I'm happy deleting a lot of code 😉

**Review by rainersigwald — APPROVED** [review]

File: src/Tasks.UnitTests/AssemblyDependency/ResolveAssemblyReferenceCacheSerialization.cs [hunk 11]
```diff
@@ -42,178 +45,63 @@ public void RoundTripEmptyState()
...
-                }
-
-                var deserialized = SystemState.DeserializeCacheByTranslator(_rarCacheFile, _taskLoggingHelper);
-
-                deserialized.ShouldBeNull();
-            }
-        }
-
-        [Fact]
-        public void WrongFileVersion()
```
[the remark this answers is not in the document]
Comment (Forgind) [reply src/Tasks.UnitTests/AssemblyDependency/ResolveAssemblyReferenceCacheSerialization.cs:74]: I can reinstate it. To be honest, in retrospect, I'm not sure why I deleted it in the first place.

File: src/Shared/BinaryTranslator.cs [hunk 12]
```diff
@@ -660,12 +660,31 @@ public void TranslateDictionary<T>(ref Dictionary<string, T> dictionary, IEquali
...
+                }
+
+                int count = _reader.ReadInt32();
+                dictionary = new(count, comparer);
+                string key = string.Empty;
+                DateTime val = DateTime.MinValue;
+                for (int i = 0; i < count; i++)
+                {
+                    Translate(ref key);
+                    Translate(ref val);
```
Comment (danmoseley) [inline src/Shared/BinaryTranslator.cs:677]: as an aside, what happens if you build, change timezones, and build again? Do we potentially under build or do we handle that?

Comment (Forgind) [reply src/Shared/BinaryTranslator.cs:677]: I believe we underbuild, but that's very unusual and perpendicular to this PR. (Does that sound right @rainersigwald?) This part is just about serializing and deserializing something.

**Review by rokonec — APPROVED** [review]

File: src/Shared/BinaryTranslator.cs (hunk 12 above)
Comment (danmoseley) [reply src/Shared/BinaryTranslator.cs:677]: Never mind - I'm pretty sure the DateTimes are all UTC so time zones can generally be ignored. DateTime barely has a concept of time zone anyway.
